### ISSUE TITLE
tidb_txn_mode don't default to pessimistic

### ISSUE THREAD

**Issue by erikthorselius** [pingcap/tidb#20990]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
docker run -d -p 4000:4000 pingcap/tidb:v4.0.8
mysql -h localhost -P 4000 -u root --protocol tcp -e "show global variables where Variable_name='tidb_txn_mode'
```
### 2. What did you expect to see? (Required)
```
+---------------+-------------+
| Variable_name | Value       |
+---------------+-------------+
| tidb_txn_mode | pessimistic |
+---------------+-------------+
```
### 3. What did you see instead (Required)
```
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| tidb_txn_mode |       |
+---------------+-------+
```
### 4. What is your TiDB version? (Required)
```
| Release Version: v4.0.8
Edition: Community
Git Commit Hash: 66ac9fc31f1733e5eb8d11891ec1b38f9c422817
Git Branch: heads/refs/tags/v4.0.8
UTC Build Time: 2020-10-30 08:21:16
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

**Comment by zyguan** [comment]
Thanks for the report, @erikthorselius , `tidb_txn_mode` defaults to pessimistic when the store is tikv, however, `docker run -d -p 4000:4000 pingcap/tidb:v4.0.8` starts a tidb with mocktikv.

**Comment by erikthorselius** [comment]
It would be nice to have a mock that have similar configuration as to the real store or maybe make it configurable from start. We have a library that checks that the database have the right configuration before it can be uses. It works well when accessing a bigger installation (read dev/prod environments) of tidb but not when just running on a developers machine.

**Comment by zyguan** [comment]
Yes, you are right, we keep it unchanged when using mocktikv because many test code rely on it. Can `set global tidb_txn_mode='pessimistic';` during provisioning resolve the problem?

**Comment by erikthorselius** [comment]
That's is how I solved it in the test's around the library right now. But it is a bit annoying because it takes around 2 seconds for the change to apply. It would be nice if it could be configured via configuration.

**Comment by zyguan** [comment]
@cfzjywxk @nullnotnil How do you think about this? Do we need to add a configuration item for it, or are there any other solutions?

**Comment by nullnotnil** [comment]
> @cfzjywxk @nullnotnil How do you think about this? Do we need to add a configuration item for it, or are there any other solutions?

I was not aware that the tests rely on optimistic, but that makes sense based on history. Without knowing how much work is involved to switch the tests, I agree with @erikthorselius that mocktikv should be using the same default. Otherwise there might be bugs in the default configuration which are not caught?

**Comment by erikthorselius** [comment]
Is it possible to take any decision? I'm just a humble user of your great software, I have a clunky workaround that I can live with. But maybe this could help other users.

**Comment by cfzjywxk** [comment]
> Is it possible to take any decision? I'm just a humble user of your great software, I have a clunky workaround that I can live with. But maybe this could help other users.
@erikthorselius 
I think it should be changed to the same, the problem that `unit test` cases may fail becasuse they depend on the old default value and we may need to refactor these cases, as @nullnotnil said. It may need to some time to change tham all.
The set statment `set global tidb_txn_mode  = 'pessimsitic'` could be used to change it currently before other statement executions.

**Comment by serbaut** [comment]
Fixed in #36578
